Topics/System_Design/Twitter_System_Design.md

```diff
@@ -0,0 +1,63 @@
+
+**users**
+| id   |      username      |  bio |
+|----------|:-------------:|------:|
+| 1 |  darth-palpatine | Unlimited Power! |
+| 2 |    anakin-skywalker   |   The Jedi are Evil |
+| 3 | obi-wan |  Hello there   |
+
+**follows**
+| follower_id   |      follows_id      |
+|----------|:-------------:|
+| 2 |  1 |
+| 2 |    3   |
+| 1 | 2 |
+
+Twitter's initial naive approach was to join the Tweet, User, and Follows tables in SQL upon a GET request from a client like so:
+```
+SELECT * FROM tweets
+JOIN users ON tweets.author_id = users.id
+JOIN follows ON follows_id = users.id
+WHERE follower_id = current_user.id;
+```
+
+You can quickly see how this approach became quite slow at over 300K QPS.
+
+## Approach 2: Caching
+As a result, Twitter implemented a fan-out caching system. They stored a [Redis list](https://redis.io/docs/data-types/lists/) for each user containing the Tweets to display on each user's home timeline. Let's illustrate this with an example. Say that Darth Palpatine is followed by Anakin Skywalker, Obi-Wan, and Padme Amidala. When Darth Palpatine posts a Tweet, Twitter's fan-out service pushes the new Tweet to the Redis cache for each user that follows Darth Palpatine (Anakin, Obi-Wan, and Padme). This way, when Anakin (for example) loads his timeline page, his Tweets are cached in Redis and can be retrieved in constant runtime (in this simplified example).
+
+The below system diagram presents a simplified version of this architecture. Note that several things are omitted for simplicity; for example the load balancer that receives the POST request from the client. Also note that the Redis Cluster is presented as one node in the diagram. In reality, a Redis Cluster (as the word "cluster" suggests) consists of lots of distributed Redis databases. In addition, Twitter (as of 2012) stores three copies of each Tweet across different Redis nodes to ensure system resilience and availability (as in a huge system like Twitter, nodes can frequently crash and become unavailable).
+
+![Twitter's Caching Approach](twitter.png)
+*A system diagram of Twitter's caching approach for home timelines.*
+
+## A Third Twist: the Hybrid Approach
+The caching approach worked for accounts with a reasonable number of followers. Assuming an average follower count of 100 and 4.6k Tweets per second (based on November 2012 data), this translated to 460k writes per second to the Redis cluster.
+
+However, for celebrity accounts with huge numbers of followers, this was not performant. A Tweet from Lebron James, who has 50 million followers as of 2023, would result in over 50 million writes at once. In this scenario, the caching approach was not feasible given Twitter's goal of delivering Tweets within 5 seconds. As a result, Twitter implemented a hybrid approach, where most new Tweets were cached (using approach 2) while Tweets from accounts with large followings were queried at runtime (like approach 1) and merged chronologically with the Tweet cache.
```

Comment:
What is the benefit to the hybrid approach for large accounts? Wouldn't the same issues exist for approach 1 with large celebrity accounts?

Reply:
Also is this hybrid approach employed at a certain amount of followers? Curious what the exact criteria may be (i.e., what if someone has 1 million followers vs 50 million?)

Reply:
To understand why the hybrid approach is better, you can broadly think about the advantages and disadvantages of approach 1 and 2.

Approach 1's complexity increases as the size of our Users, Tweets, and Followers tables gets really large. However the catch is that Twitter users don't really follow millions of users, so we only have to consider Users and Tweets.

Approach 2's complexity increases based on how many accounts follow you.

We've established that Approach 2 isn't the best with celebrities. Now let's consider the benefits of the hybrid approach over Approach 1.

We can create tables CelebrityTweets and CelebrityUsers. When a user queries their homepage, we get from the Redis cache in O(1) runtime. Then we JOIN CelebrityTweets and CelebrityUsers to get any Tweets from celebrity accounts they might follow. Note that the size of CelebrityTweets and CelebrityUsers is quite limited relatively (to our original approach).

Of course, given the popularity of Twitter, this hybrid approach might still have scalability issues (this is why scalability is a complex issue). But the point of the ultra-simple example in this article is that the hybrid approach is much better than _just_ doing Approach 1.

Do you think this explanation makes sense? I can add it to the article if you feel like it would be a good addition.

Reply:
And not sure about exact follower cutoffs. I think that's out of the scope of this article and probably a private implementation detail to Twitter.

Reply:
I think the explanation is well-written; worth adding I think.